src/lib/storage/dictionary_segment.cpp

```diff
@@ -0,0 +1,133 @@
+#include <unordered_set>
+
+#include "dictionary_segment.hpp"
+#include "type_cast.hpp"
+
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+template <typename T>
+DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  for (auto index = ChunkOffset{0}; index < abstract_segment->size(); index++) {
+    const auto& value = type_cast<T>(abstract_segment->operator[](index));
+    _dictionary.push_back(value);
+  }
+
+  // use standard library functions to remove duplicate values from dictionary
+  // sorted dictionary also allows us to be more efficient in element lookup
+  std::sort(begin(_dictionary), end(_dictionary));
+  const auto& end_it = std::unique(begin(_dictionary), end(_dictionary));
```

Comment:
Interesting approach. We created a separate std::set to collect the distinct values and then copied the values into a final vector. You are allocating less memory by reusing the existing vector space. Cool!